cloud/cloud_env_impl.cc

```diff
@@ -1796,47 +1803,113 @@ Status CloudEnvImpl::RollNewEpoch(const std::string& local_dbname) {
   }
   // roll new epoch
   auto newEpoch = generateNewEpochId();
-  GetCloudManifest()->AddEpoch(maxFileNumber, newEpoch);
-  GetCloudManifest()->Finalize();
+  // To make sure `RollNewEpoch` is backwards compatible, we don't change
+  // the cookie when applying CM delta
+  auto newCookie = cloud_env_options.cookie_on_open;
+
+  st = ApplyLocalCloudManifestDelta(
+      local_dbname,
+      newCookie,
+      CloudManifestDelta{maxFileNumber, newEpoch});
+  if (!st.ok()) {
+    return st;
+  }
+
+  // TODO(igor): Compact cloud manifest by looking at live files in the database
+  // and removing epochs that don't contain any live files.
+
+  if (HasDestBucket()) {
+    st = UploadLocalCloudManifestAndManifest(local_dbname, newCookie);
+    if (!st.ok()) {
+      return st;
+    }
+  }
+  return Status::OK();
+}
+
+Status CloudEnvImpl::UploadLocalCloudManifestAndManifest(
+    const std::string& local_dbname, const std::string& cookie) {
+  assert(HasDestBucket());
+
+  std::string current_epoch = cloud_manifest_->GetCurrentEpoch();
+  // We have to upload the manifest file first. Otherwise, if the process
+  // crashed in the middle, we'll endup with a CLOUDMANIFEST file pointing to
+  // MANIFEST file which doesn't exist in s3
+  auto st = GetStorageProvider()->PutCloudObject(
+      ManifestFileWithEpoch(local_dbname, current_epoch), GetDestBucketName(),
+      ManifestFileWithEpoch(GetDestObjectPath(), current_epoch));
+  if (!st.ok()) {
+    return st;
+  }
+
+  TEST_SYNC_POINT_CALLBACK(
+      "CloudEnvImpl::UploadLocalCloudManifestAndManifest:AfterUploadManifest",
+      &st);
+  if (!st.ok()) {
+    return st;
+  }
+
+  return UploadLocalCloudManifest(local_dbname, cookie);
+}
+
+Status CloudEnvImpl::UploadLocalCloudManifest(const std::string& local_dbname,
+                                              const std::string& cookie) {
+   // upload the cloud manifest file corresponds to cookie (i.e., CLOUDMANIFEST-cookie)
```

Comment:
This is a public function. We should fail if there is no DestBucket in case of the caller's error.